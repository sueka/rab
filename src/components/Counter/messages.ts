import { defineMessages } from 'react-intl'

export default defineMessages({
  reset: {
    id: 'src.components.Counter.reset',
    defaultMessage: 'reset',
  },
  increment: {
    id: 'src.components.Counter.increment',
    defaultMessage: '+',
  },
  decrement: {
    id: 'src.components.Counter.decrement',
    defaultMessage: '-',
  },
  incrementIfOdd: {
    id: 'src.components.Counter.incrementIfOdd',
    defaultMessage: '+ if odd',
  },
  willIncrementInOneSecond: {
    id: 'src.components.Counter.willIncrementInOneSecond',
    defaultMessage: '+ in 1 second',
  },
})
