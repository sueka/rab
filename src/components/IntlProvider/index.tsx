import * as React from 'react'
import { IntlProvider as OriginalIntlProvider } from 'react-intl'

export type StateProps = Pick<OriginalIntlProvider.Props, 'locale' | 'formats' | 'messages'>

type Props =
  & StateProps

// TODO: intl context でない要素を re-render しないようにする。 https://github.com/formatjs/react-intl/issues/234#issuecomment-163366518 によると現時点では難しいらしい。
// cf. https://github.com/formatjs/react-intl/issues/371#issuecomment-275703796
const IntlProvider: React.FunctionComponent<Props> = (props) => (
  <OriginalIntlProvider key={ props.locale } { ...props } />
)

export default IntlProvider
