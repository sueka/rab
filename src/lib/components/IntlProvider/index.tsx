import React from 'react'
import { IntlConfig, IntlProvider as OriginalIntlProvider } from 'react-intl'

import IntlProviderContext from '~/lib/contexts/IntlProviderContext'
import { Tag } from '~/lib/languageNameSolver'

type DefaultIntlConfig = Pick<IntlConfig, 'formats' | 'messages' | 'timeZone' | 'textComponent' | 'defaultLocale' | 'defaultFormats' | 'onError'>

type OriginalIntlProviderProps = Alt.Omit<IntlConfig, keyof DefaultIntlConfig> & Partial<DefaultIntlConfig>

export interface OwnProps extends Alt.Omit<OriginalIntlProviderProps, 'locale' | 'formats' | 'messages'> {
  availableLocales: Tag[]
}

export type StateProps = Pick<OriginalIntlProviderProps, 'locale' | 'formats' | 'messages'>

type Props =
  & OwnProps
  & StateProps

// NOTE: key が無い場合、 FormattedMessage 等は re-render されるが、 useIntl の結果は更新されない。
// TODO: intl context でない要素を re-render しないようにする。 https://github.com/formatjs/react-intl/issues/234#issuecomment-163366518 によると現時点では難しいらしい。
// cf. https://github.com/formatjs/react-intl/issues/371#issuecomment-275703796
const IntlProvider: React.FunctionComponent<Props> = ({ availableLocales, ...props }) => (
  <IntlProviderContext.Provider value={ { availableLocales } }>
    <OriginalIntlProvider key={ props.locale } textComponent={ React.Fragment } { ...props } />
  </IntlProviderContext.Provider>
)

export default IntlProvider
