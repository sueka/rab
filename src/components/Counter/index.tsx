import * as React from 'react'
import { FormattedMessage } from 'react-intl'

import Button from '@material-ui/core/Button'

import messages from './messages'

export interface StateProps {
  value: number
}

export interface DispatchProps {
  reset(): void
  increment(): void
  decrement(): void
  incrementIfOdd(): void
  incrementAsync(ms: number): void
}

type Props =
  & StateProps
  & DispatchProps

const Counter: React.FunctionComponent<Props> = ({ value, reset, increment, decrement, incrementIfOdd, incrementAsync }) => {
  const handleIncrementAsync = React.useCallback(() => {
    incrementAsync(1000)
  }, [])

  return (
    <div>
      { value }
      <Button onClick={ reset }><FormattedMessage { ...messages.reset } /></Button>
      <Button onClick={ increment }><FormattedMessage { ...messages.increment } /></Button>
      <Button onClick={ decrement }><FormattedMessage { ...messages.decrement } /></Button>
      <Button onClick={ incrementIfOdd } data-testid="incrementIfOddButton"><FormattedMessage { ...messages.incrementIfOdd } /></Button>
      <Button onClick={ handleIncrementAsync } data-testid="incrementAsyncButton"><FormattedMessage { ...messages.willIncrementInOneSecond } /></Button>
    </div>
  )
}

export default Counter
