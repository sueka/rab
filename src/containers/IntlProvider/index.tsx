import { connect } from 'react-redux'

import { State } from '../../redux'
import IntlProvider, { StateProps } from '../../components/IntlProvider'

const mapStateToProps = ({ localeSelector: { locale, messages } }: State): StateProps => ({
  locale,
  messages,
})

export default connect(mapStateToProps)(IntlProvider)
