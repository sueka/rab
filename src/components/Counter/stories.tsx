import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withIntlProvider, withProvider } from '~/storybook/decorators'
import Counter from '.'

storiesOf('Counter', module)
.addDecorator(withProvider)
.addDecorator(withIntlProvider)
.add('default', () => (<Counter />))
