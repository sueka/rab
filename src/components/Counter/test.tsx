import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import IntlProvider from '~/components/IntlProvider'
import Counter from '.'

test('Counter', () => {
  const reset = jest.fn()
  const increment = jest.fn()
  const decrement = jest.fn()
  const incrementIfOdd = jest.fn()
  const incrementAsync = jest.fn()

  const { container, getByTestId } = render(
    <IntlProvider locale="en">
      <Counter
        value={ 0 }
        reset={ reset }
        increment={ increment }
        decrement={ decrement }
        incrementIfOdd={ incrementIfOdd }
        incrementAsync={ incrementAsync }
      />
    </IntlProvider>
  )

  expect(container.firstChild).toMatchSnapshot()

  fireEvent.click(getByTestId('incrementIfOddButton'))
  expect(incrementIfOdd).toHaveBeenCalledTimes(1)

  fireEvent.click(getByTestId('incrementAsyncButton'))
  expect(incrementAsync).toHaveBeenCalledTimes(1)
})
