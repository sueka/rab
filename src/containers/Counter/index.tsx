import { connect } from 'react-redux'

import { State } from '../../redux'
import { increment, decrement, incrementIfOdd, incrementAsync } from '../../redux/modules/counter'
import Counter, { StateProps, DispatchProps } from '../../components/Counter'

const mapStateToProps = ({ counter: { count } }: State): StateProps => ({
  value: count,
})

const mapDispatchToProps: DispatchProps = {
  increment,
  decrement,
  incrementIfOdd,
  incrementAsync,
}

export default connect(mapStateToProps, mapDispatchToProps)(Counter)
