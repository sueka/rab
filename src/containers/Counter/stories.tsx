import * as React from 'react'
import { storiesOf } from '@storybook/react'

import { withProvider } from '../../storybook/decorators'
import Counter from '.'

storiesOf('Counter', module)
  .addDecorator(withProvider)
  .add('default', () => (<Counter />))
