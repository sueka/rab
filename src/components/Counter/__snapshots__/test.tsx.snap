// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter 1`] = `
<div>
  <span
    class="MuiTypography-root MuiTypography-body1"
  >
    0
  </span>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      reset
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      +
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      -
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    data-testid="incrementIfOddButton"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      + if odd
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    data-testid="incrementAsyncButton"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      + in 1 second
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
